File: communication/rabbitmq/simple-demo/go-server/consumer/listener.go
package consumer

import (
	"encoding/json"
	"log"

	amqp "github.com/rabbitmq/amqp091-go"

	"github.com/seungyeop-lee/poc/communication/rabbitmq/simple-demo/go-server/common"
	"github.com/seungyeop-lee/poc/communication/rabbitmq/simple-demo/go-server/config"
)

type Listener struct {
	rabbit *config.RabbitMQ
}

func NewListener(rabbit *config.RabbitMQ) *Listener {
	return &Listener{rabbit: rabbit}
}

func (l *Listener) StartConsuming() error {
	err := l.setupExchangeAndQueue()
	if err != nil {
		return err
	}

	msgs, err := l.rabbit.Channel.Consume(
		common.QueueName,
		"",
		true,
		false,
		false,
		false,
		nil,
	)
	if err != nil {
		return err
	}

	go l.handleMessages(msgs)
	return nil
}

func (l *Listener) setupExchangeAndQueue() error {
	err := l.rabbit.Channel.ExchangeDeclare(
		common.ExchangeName,
		"topic",
		true,
		false,
		false,
		false,
		nil,
	)
	if err != nil {
		return err
	}

	q, err := l.rabbit.Channel.QueueDeclare(
		common.QueueName,
		true,
		false,
		false,
		false,
		nil,
	)
	if err != nil {
		return err
	}

	return l.rabbit.Channel.QueueBind(
		q.Name,
		"poc.communication.rabbitmq.#",
		common.ExchangeName,
		false,
		nil,
	)
}

func (l *Listener) handleMessages(msgs <-chan amqp.Delivery) {
	for msg := range msgs {
		log.Printf("Received ContentType: %s", msg.ContentType)

		var payload common.PayloadMessage
		_ = json.Unmarshal(msg.Body, &payload)
		log.Printf("Received Message: %+v\n", payload)
		log.Printf("Received Message Data: %s\n", string(payload.Data))
		log.Printf("Received RoutingKey: %s\n", msg.RoutingKey)
	}
}

File: communication/rabbitmq/simple-demo/python-server/python_server/consumer/listener.py
import base64
import json
import logging

import aio_pika

from python_server.common.constants import EXCHANGE_NAME, QUEUE_NAME
from python_server.common.types import PayloadMessage
from python_server.config.rabbitmq import RabbitMQ


class Listener:
    def __init__(self, rabbit: RabbitMQ):
        self.rabbit = rabbit

    async def start_consuming(self):
        queue = await self.setup_exchange_and_queue()
        await queue.consume(handle_message)

    async def setup_exchange_and_queue(self):
        exchange = await self.rabbit.channel.declare_exchange(
            name=EXCHANGE_NAME,
            type=aio_pika.ExchangeType.TOPIC,
            durable=True,
        )

        queue = await self.rabbit.channel.declare_queue(
            QUEUE_NAME,
            durable=True,
        )

        await queue.bind(
            exchange=exchange,
            routing_key="poc.communication.rabbitmq.#",
        )
        return queue


async def handle_message(message: aio_pika.IncomingMessage):
    async with message.process():
        logging.info(f"Received ContentType: {message.content_type}")
        body = json.loads(message.body.decode())
        payload = PayloadMessage(**body)
        logging.info(f"Received Message: {payload}")
        logging.info(f"Received Message Data: {base64.b64decode(payload.data).decode('utf-8')}")
        logging.info(f"Received RoutingKey: {message.routing_key}")
